internal/exec/util/selabeling.c

```diff
@@ -0,0 +1,160 @@
+// vim:set sw=2 sts=2 et: tw=92
+//
+// Copyright 2019 Red Hat, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#define _GNU_SOURCE
+#include <assert.h>
+#include <errno.h>
+#include <limits.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdarg.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <sys/xattr.h>
+#include <unistd.h>
+
+#include <selinux/selinux.h>
+#include <selinux/label.h>
+
+#include "selabeling.h"
+
+/* The main reason we do this from C here instead of through cgo (apart from the
+ * latter being painful to work with) is for error-handling: we're using the
+ * APIs which set errno, which we simply cannot access from Go:
+ * - https://github.com/golang/go/issues/1360
+ * - https://codereview.appspot.com/3880041/
+ *
+ * Relatedly, we construct a good error string here instead of returning a
+ * result_t style enum like in `blkid.c`. We could combine the two in the future
+ * if we need the Go side to special-case an error, though today these errors
+ * are not really actionable.
+ */
+
+struct selabeling_context {
+    char *root;
+    struct selabel_handle *handle;
+};
+
+static char*
+malloc_vprintf(const char *fmt, va_list args)
+{
+  /* see vsnprintf(3) EXAMPLE */
+  va_list argsc;
+  va_copy(argsc, args);
+  int n = vsnprintf(NULL, 0, fmt, argsc);
+  va_end(argsc);
+  assert(n > 0);
+  n++; /* for final NUL */
+  char *s = malloc(n);
+  assert(vsnprintf(s, n, fmt, args) == n-1);
+  return s;
+}
+
+static char*
+malloc_sprintf(const char *fmt, ...)
+{
+  va_list args;
+  va_start (args, fmt);
+  char *s = malloc_vprintf(fmt, args);
+  va_end (args);
+  return s;
+}
+
+/* see similar UX function glnx_throw_errno_prefix() in libglnx */
+static int
+throw_errno_prefix(char **error, const char *fmt, ...)
+{
+  if (!error)
+    return -1;
+
+  int errsv = errno;
+  va_list args;
+  va_start (args, fmt);
+
+  char *prefix = malloc_vprintf(fmt, args);
+  char buf[1024]; /* see NOTES in strerror(3) */
+  *error = malloc_sprintf("%s: %s", prefix, strerror_r(errsv, buf, sizeof(buf)));
```

Comment:
glibc ships `asprintf`, why not just use that?

Reply:
Ahh yes, that's much better! I somehow had the impression that `asprintf` allocated on the stack, but that's just `alloca`.